File: frontend/src/utils/habitSchedule.ts
import type { Habit, HabitEntry, WeekdayKey } from "../types/habit";
import { formatLocalDate, formatPreferredTime } from "./formatDate";
import { weekdayKeys } from "./habitForm";

const weekdayLabels: Record<WeekdayKey, string> = {
  mon: "Пн",
  tue: "Вт",
  wed: "Ср",
  thu: "Чт",
  fri: "Пт",
  sat: "Сб",
  sun: "Вс"
};

export type HabitUnavailableReason = "not-scheduled" | "first-after-preferred-time";

export interface HabitScheduleAvailability {
  isScheduledToday: boolean;
  isAvailableToday: boolean;
  isPastPreferredTime: boolean;
  reason?: HabitUnavailableReason;
}

export interface ScheduledOccurrence {
  date: Date;
  weekday: WeekdayKey;
  time: string | null;
}

export type HabitPeriodState =
  | "not_due"
  | "available_to_complete"
  | "completed"
  | "missed"
  | "locked";

export interface HabitCurrentPeriodState {
  state: HabitPeriodState;
  availability: HabitScheduleAvailability;
  canComplete: boolean;
  canSkip: boolean;
  isAlreadyCounted: boolean;
}

function getWeekdayKey(date: Date): WeekdayKey {
  return weekdayKeys[(date.getDay() + 6) % 7];
}

function getEffectiveScheduleDays(habit: Habit): WeekdayKey[] {
  if (habit.schedule_days.length > 0) {
    return habit.schedule_days;
  }

  return habit.frequency_type === "daily" ? weekdayKeys : [];
}

function getTimeMinutes(value: string | null): number | undefined {
  if (!value) {
    return undefined;
  }

  const [hours, minutes] = value.split(":");
  const parsedHours = Number(hours);
  const parsedMinutes = Number(minutes);
  if (!Number.isInteger(parsedHours) || !Number.isInteger(parsedMinutes)) {
    return undefined;
  }

  return parsedHours * 60 + parsedMinutes;
}

function getDateMinutes(date: Date): number {
  return date.getHours() * 60 + date.getMinutes();
}

function startOfLocalDay(date: Date): Date {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
}

function parseAsUTC(isoString: string): Date {
  // Backend stores naive UTC datetimes without 'Z'. Force UTC interpretation so that
  // getFullYear/getMonth/getDate return LOCAL calendar values, not UTC calendar values.
  const hasZone = isoString.endsWith("Z") || /[+-]\d{2}:?\d{2}$/.test(isoString);
  return new Date(hasZone ? isoString : isoString + "Z");
}

function shouldDeferFirstOccurrence(habit: Habit, now: Date, hasEntries: boolean) {
  const preferredMinutes = getTimeMinutes(habit.preferred_time);
  if (hasEntries || preferredMinutes === undefined) return false;

  const createdLocalDate = formatLocalDate(parseAsUTC(habit.created_at));
  return createdLocalDate === formatLocalDate(now) && getDateMinutes(now) > preferredMinutes;
}

export function getHabitScheduleAvailability(
  habit: Habit,
  now: Date,
  hasEntries = false
): HabitScheduleAvailability {
  const isScheduledToday = getEffectiveScheduleDays(habit).includes(getWeekdayKey(now));
  if (!isScheduledToday) {
    return {
      isScheduledToday,
      isAvailableToday: false,
      isPastPreferredTime: false,
      reason: "not-scheduled"
    };
  }

  const preferredMinutes = getTimeMinutes(habit.preferred_time);
  const isPastPreferredTime =
    preferredMinutes !== undefined && getDateMinutes(now) > preferredMinutes;

  if (shouldDeferFirstOccurrence(habit, now, hasEntries)) {
    return {
      isScheduledToday,
      isAvailableToday: false,
      isPastPreferredTime,
      reason: "first-after-preferred-time"
    };
  }

  return {
    isScheduledToday,
    isAvailableToday: true,
    isPastPreferredTime
  };
}

export function getHabitCurrentPeriodState(
  habit: Habit,
  now: Date,
  hasEntries: boolean,
  todayEntry?: HabitEntry
): HabitCurrentPeriodState {
  const availability = getHabitScheduleAvailability(habit, now, hasEntries);
  const isCompleted =
    todayEntry?.status === "completed" || todayEntry?.status === "recovery_completed";

  if (isCompleted) {
    return {
      state: "completed",
      availability,
      canComplete: false,
      canSkip: false,
      isAlreadyCounted: true
    };
  }

  if (todayEntry?.status === "missed") {
    return {
      state: "missed",
      availability,
      canComplete: false,
      canSkip: false,
      isAlreadyCounted: true
    };
  }

  if (!availability.isScheduledToday || !availability.isAvailableToday) {
    return {
      state: "not_due",
      availability,
      canComplete: false,
      canSkip: false,
      isAlreadyCounted: false
    };
  }

  return {
    state: "available_to_complete",
    availability,
    canComplete: true,
    canSkip: false,
    isAlreadyCounted: false
  };
}

export function getNextScheduledOccurrence(
  habit: Habit,
  now: Date,
  startDayOffset = 0
): ScheduledOccurrence | undefined {
  const scheduleDays = new Set(getEffectiveScheduleDays(habit));
  if (scheduleDays.size === 0) {
    return undefined;
  }

  for (let dayOffset = startDayOffset; dayOffset <= startDayOffset + 7; dayOffset += 1) {
    const date = startOfLocalDay(now);
    date.setDate(date.getDate() + dayOffset);
    const weekday = getWeekdayKey(date);

    if (!scheduleDays.has(weekday)) {
      continue;
    }

    return {
      date,
      weekday,
      time: habit.preferred_time
    };
  }

  return undefined;
}

function formatScheduledOccurrence(
  label: string,
  occurrence: ScheduledOccurrence | undefined
): string {
  if (!occurrence) {
    return `${label} не запланировано`;
  }

  const dateLabel = occurrence.date.toLocaleDateString("ru-RU", {
    day: "numeric",
    month: "long"
  });

  return `${label}: ${weekdayLabels[occurrence.weekday]}, ${dateLabel}, ${formatPreferredTime(occurrence.time)}`;
}

export function formatNextScheduledOccurrence(occurrence: ScheduledOccurrence | undefined): string {
  return formatScheduledOccurrence("Следующий раз", occurrence);
}

export function formatFirstScheduledOccurrence(occurrence: ScheduledOccurrence | undefined): string {
  return formatScheduledOccurrence("Первый раз", occurrence);
}

File: backend/app/services/habit_schedule.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional

WEEKDAY_KEYS = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")


def _normalize_schedule_day(day: Any) -> Optional[str]:
    if isinstance(day, int) and 0 <= day <= 6:
        return WEEKDAY_KEYS[day]
    if isinstance(day, str):
        value = day.strip().lower()
        if value.isdigit():
            return _normalize_schedule_day(int(value))
        if value in WEEKDAY_KEYS:
            return value
    return None


def get_schedule_days(habit: Any) -> set[str]:
    days = {
        normalized
        for normalized in (_normalize_schedule_day(day) for day in (habit.schedule_days or []))
        if normalized is not None
    }
    if days:
        return days
    if getattr(habit, "frequency_type", None) == "daily":
        return set(WEEKDAY_KEYS)
    return set()


def is_habit_scheduled_on(habit: Any, target_date: date) -> bool:
    return WEEKDAY_KEYS[target_date.weekday()] in get_schedule_days(habit)


def get_next_scheduled_date(
    habit: Any,
    start_date: date,
    start_day_offset: int = 0,
) -> Optional[date]:
    schedule_days = get_schedule_days(habit)
    if not schedule_days:
        return None

    for day_offset in range(start_day_offset, start_day_offset + 8):
        target_date = start_date.fromordinal(start_date.toordinal() + day_offset)
        if WEEKDAY_KEYS[target_date.weekday()] in schedule_days:
            return target_date

    return None


def is_first_occurrence_deferred(habit: Any, now: datetime) -> bool:
    created_at = getattr(habit, "created_at", None)
    preferred_time = getattr(habit, "preferred_time", None)
    entries = getattr(habit, "entries", [])

    return bool(
        isinstance(created_at, datetime)
        and preferred_time is not None
        and created_at.date() == now.date()
        and now.time() > preferred_time
        and not entries
    )


def is_habit_available_at(habit: Any, now: datetime) -> bool:
    return is_habit_scheduled_on(habit, now.date()) and not is_first_occurrence_deferred(
        habit,
        now,
    )


def get_today_unavailability_detail(habit: Any, now: datetime) -> Optional[str]:
    if not is_habit_scheduled_on(habit, now.date()):
        return "Привычка не запланирована на сегодня"
    if is_first_occurrence_deferred(habit, now):
        return "Первый шаг перенесен на следующий запланированный день"
    return None
